src/box/box.cc

```diff
@@ -4662,6 +4662,72 @@ box_read_ffi_enable(void)
 		box_read_ffi_is_disabled = false;
 }
 
+API_EXPORT int64_t
+box_advance_max_id(void)
+{
+	/* ["max_id"] */
+	static const char key[] = {
+		(char)0x91, /* MsgPack array(1) */
+		(char)0xa6, /* MsgPack string(6) */
+		'm', 'a', 'x', '_', 'i', 'd'
+	};
+	/* [["+", 1, 1]]*/
+	static const char ops[] = {
+		(char)0x91, /* MsgPack array(1) */
+		(char)0x93, /* MsgPack array(3) */
+		(char)0xa1, /* MsgPack string(1) */
+		'+',
+		1,          /* MsgPack int(1) */
+		1           /* MsgPack int(1) */
+	};
+
+	uint64_t max_id;
+	struct tuple *res = NULL;
+	struct request request;
+	memset(&request, 0, sizeof(request));
+	request.tuple = ops;
+	request.tuple_end = ops + sizeof(ops);
+	request.key = key;
+	request.key_end = key + sizeof(key);
+	request.type = IPROTO_UPDATE;
+	request.space_id = BOX_SCHEMA_ID;
+
+	if (box_process1(&request, &res) != 0)
+		return -1;
+	if (res != NULL) {
+		if (tuple_field_u64(res, 1, &max_id) != 0)
+			return -1;
+	} else {
+		max_id = BOX_SYSTEM_ID_MAX;
+	}
+	/*
+	 * Use the first available id greater than BOX_SYSTEM_ID_MAX if max_id
+	 * is overflowed.
+	 */
+	if (max_id > BOX_SPACE_MAX)
+		max_id = BOX_SYSTEM_ID_MAX;
+	bool max_id_is_system = false;
+	if (max_id <= BOX_SYSTEM_ID_MAX) {
+		max_id = BOX_SYSTEM_ID_MAX + 1;
+		max_id_is_system = true;
+	}
+	uint32_t available_max_id = space_available_id(max_id);
+	/** Try again if overflowed. */
+	if (available_max_id > BOX_SPACE_MAX && !max_id_is_system) {
+		max_id = BOX_SYSTEM_ID_MAX + 1;
+		available_max_id = space_available_id(max_id);
```

Comment:
AFAIU the function may fail to find a vacant id, but you never check that.

Reply:
The logic is if we overflowed the second time - overflowed value is returned and the space creation will fail with an error. I did so because the error is more informative - in this function we don't know the new space name.

Reply:
What kind of error? Is it tested? IMO an explicit error indicating that we ran out of space ids would be better.

Reply:
Discussed f2f, use panic here